src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -437,27 +492,39 @@ end = struct
 
   (** Handle the initial Num_accounts message, starting the main syncing
       process. *)
-  let handle_num_accounts t n content_hash =
+  let handle_num_accounts :
+      'a t -> int -> Hash.t -> [`Success | `Hash_mismatch of Hash.t * Hash.t] =
+   fun t n content_hash ->
     let rh = Root_hash.to_hash (desired_root_exn t) in
     let height = Int.ceil_log2 n in
     (* FIXME: bug when height=0 https://github.com/o1-labs/nanobit/issues/365 *)
-    if not (Hash.equal (complete_with_empties content_hash height MT.depth) rh)
-    then failwith "reported content hash doesn't match desired root hash!" ;
-    (* TODO: punish *)
-    MT.make_space_for t.tree n ;
-    Addr.Table.clear t.waiting_parents ;
-    (* We should use this information to set the empty account slots empty and
-       start syncing at the content root. See #1972. *)
-    Addr.Table.clear t.waiting_content ;
-    handle_node t (Addr.root ()) rh
+    let actual = complete_with_empties content_hash height MT.depth in
+    if Hash.equal actual rh then (
+      MT.make_space_for t.tree n ;
+      Addr.Table.clear t.waiting_parents ;
+      (* We should use this information to set the empty account slots empty and
+         start syncing at the content root. See #1972. *)
+      Addr.Table.clear t.waiting_content ;
+      handle_node t (Addr.root ()) rh ;
+      `Success )
+    else `Hash_mismatch (rh, actual)
 
   let main_loop t =
-    let handle_answer (root_hash, query, env) =
+    let handle_answer :
+           Root_hash.t
+           * Addr.t Query.t
+           * (Hash.t, Account.t) Answer.t Envelope.Incoming.t
+        -> unit Deferred.t =
+     fun (root_hash, query, env) ->
+      (* NOTE: think about synchronization here. This is deferred now, so
```

Comment:
I'm not worried about races to the underlying ledger. A long-standing invariant of `syncable_ledger` is that it is the only writer to the destination ledger.

Reply:
Which should probably be documented somewhere...

Reply:
(especially since the current impl won't fall over and die immediately in the case of conflicting ownership)

Reply:
I meant races between responses to different queries, *inside* syncable ledger.